File: LC100/lc71.py
"""
Given an absolute path for a file (Unix-style), simplify it.

For example,
path = "/home/", => "/home"
path = "/a/./b/../../c/", => "/c"

Corner Cases:
Did you consider the case where path = "/../"?
In this case, you should return "/".
Another corner case is the path might contain multiple slashes '/' together,
 such as "/home//foo/".
In this case, you should ignore redundant slashes and return "/home/foo".
"""


def simplify(path):
    dirs = [p for p in path.split("/") if p != "" and p != "."]
    stack = []
    for p in dirs:
        if p == "..":
            if stack:
                stack.pop()
        else:
            stack.append(p)

    return "/" + "/".join(stack)


def test():
    path_list = [
        "",
        "/",
        "/a/b/./../c/",
        "/home/a/../",
        "//",
        "/../"
    ]
    for path in path_list:
        print(simplify(path))


test()

File: LC100/lc71.js
function simplify(path) {
    var dirs, p, stack = [];
    dirs = path.split("/");
    for (p of dirs) {
        if (p === "..") {
            if (stack.length > 0) {
                stack.pop();
            }
        } else if (p === "." || p === "") {
            
        } else {
            stack.push(p);
        }
    }
    return "/" + stack.join("/");
}


function test() {
    var path_list = [
        "/a/b/./../c/",
        "/home/a/../",
        "",
        "/",
        "//",
        "/../"
    ];
    for (var path of path_list) {
        console.log(simplify(path));
    }
}


test();
